crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -3431,6 +3458,143 @@ impl<'db> TypeInferenceBuilder<'db> {
         self.types.shrink_to_fit();
         self.types
     }
+
+    fn infer_subscript_type_expression(
+        &mut self,
+        subscript: &ruff_python_ast::ExprSubscript,
+    ) -> Type<'db> {
+        let ast::ExprSubscript {
+            range: _,
+            value,
+            slice,
+            ctx: _,
+        } = subscript;
+
+        let value_ty = self.infer_type_expression(value);
+
+        // Handling of Special Forms
+        match value_ty {
+            Type::Instance(instance)
+                if instance.is_known_class(self.db, KnownClass::SpecialForm) =>
+            {
+                self.infer_parameterized_known_instance_type_expression(instance, slice)
+            }
+            value_ty => {
+                let value_node = value.as_ref();
+                let slice_ty = self.infer_expression(slice);
+                // TODO: currently the logic to get the type of type of a subscript in type
```

Comment:
I needed to keep this here because we have a test case that checks we emit unsubscriptable error in type annotations so I kept it here with a todo to not break that test.